spark/src/test/scala/org/apache/spark/sql/delta/DeltaInsertIntoTest.scala

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta
+
+import org.apache.spark.sql.delta.test.DeltaSQLCommandTest
+
+import org.apache.spark.{DebugFilesystem, SparkThrowable}
+import org.apache.spark.sql.{DataFrame, QueryTest, SaveMode}
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.functions.{col, lit}
+import org.apache.spark.sql.types.StructType
+
+/**
+ * There are **many** different ways to run an insert:
+ * - Using SQL, the dataframe v1 and v2 APIs or the streaming API.
+ * - Append vs. Overwrite / Partition overwrite.
+ * - Position-based vs. name-based resolution.
+ *
+ * Each take a unique path through analysis. The abstractions below captures these different
+ * inserts to allow more easily running tests with all or a subset of them.
+ */
+trait DeltaInsertIntoTest extends QueryTest with DeltaDMLTestUtils with DeltaSQLCommandTest {
```

Comment:
This is moved from `TypeWideningInsertSchemaEvolutionSuite`.

`StreamingInsert` was added in the process to cover streaming writes, and the test runner method `testInserts` was updated to make it slightly more expressive